### PR TITLE
hw-manager: reimplement PCIe bifurcation configuration

### PR DESCRIPTION
**Pull request by alatarum**
Change options format to be more flexible in configure port halves.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

### CODE REVIEW

**Review by nest1ing — DISMISSED** [review]

File: src/hw/pcie_cfg.cpp [hunk 1]
```diff
@@ -27,6 +27,17 @@ static constexpr uint8_t pcieBifurcateXXX8XXX8 = 3;
...
+static constexpr uint8_t bmcPcieBifurcateX4X4X4X4 = 0x44;
+static constexpr uint8_t bmcPcieBifurcateX4X4XXX8 = 0x48;
+static constexpr uint8_t bmcPcieBifurcateXXX8X4X4 = 0x84;
+static constexpr uint8_t bmcPcieBifurcateXXX8XXX8 = 0x88;
+static constexpr uint8_t bmcPcieBifurcateXXXXXX16 = 0x16;
+static constexpr uint8_t bmcPcieBifurcate____X4X4 = 0xF4;
+static constexpr uint8_t bmcPcieBifurcate____XXX8 = 0xF8;
+static constexpr uint8_t bmcPcieBifurcateX4X4____ = 0x4F;
+static constexpr uint8_t bmcPcieBifurcateXXX8____ = 0x8F;
+static constexpr uint8_t bmcPcieBifurcateXXXXXXXX = 0xFF;
```
Comment (AlexanderAmelkin) [inline src/hw/pcie_cfg.cpp]: Looks a bit inconsistent. If you use underscores above for 'unchanged' bifurcation, then why use `X` here ?
Maybe call this one `bmcPcieBifurcateUnchanged` ? Or call the above ones `...BifurcateXXXXX4X4`, `...BifurcateX4X4XXXX` and so on?

File: src/hw/pcie_cfg.cpp [hunk 2]
```diff
@@ -83,6 +94,37 @@ bool pcieCfg::addBifurcationConfig(const int& socket,
                           entry("VALUE=%s", slots.c_str()));
     }
 
+    // FIXME BBMC-1020: we hardcode RADUNI address for now since currently we
+    // only support Rx20 Gen1 and it is said that RADUNI will always connected
+    // to J45 (B1)
```
Comment (AlexanderAmelkin) [inline src/hw/pcie_cfg.cpp]: I think "to CPU0 port PE2 via J45 (B1)" would be more clear

File: src/hw/pcie_cfg.cpp [hunk 3]
```diff
@@ -83,6 +94,37 @@ bool pcieCfg::addBifurcationConfig(const int& socket,
                           entry("VALUE=%s", slots.c_str()));
     }
 
+    // FIXME BBMC-1020: we hardcode RADUNI address for now since currently we
+    // only support Rx20 Gen1 and it is said that RADUNI will always connected
```
Comment (AlexanderAmelkin) [inline src/hw/pcie_cfg.cpp]: "will always connect" or "will always be connected"

File: src/hw/pcie_cfg.cpp [hunk 4]
```diff
@@ -97,32 +139,16 @@ bool pcieCfg::addBifurcationConfig(const int& socket,
...
-        else if ((old_value == pcieBifurcateXXX8XXX8) &&
-                 ((value == pcieBifurcateX4X4XXX8) ||
-                  (value == pcieBifurcateXXX8X4X4)))
+        else if (((old_value & 0x0f) == 0x0f) && ((value & 0xf0) == 0xf0))
         {
-            // We want to split our half into 'x4x4', while other component uses
-            // it's half as 'x8' - set our value
-            bifurcationConfig[instance] = value;
+            // merge High half of old value with Low half of current value
+            bifurcationConfig[instance] = (old_value & 0xf0) | (value & 0x0f);
```
Comment (AlexanderAmelkin) [inline src/hw/pcie_cfg.cpp:156]: What if the new value is `bmcPcieBifurcateXXXXXXXX` that is `0xFF` ? Will that work?

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: src/hw/pcie_cfg.cpp (hunk 1 above)
Comment (alatarum) [reply src/hw/pcie_cfg.cpp]: AFAIK it is not "Unchanged", it is "Disabled". But can change, sure.

File: src/hw/pcie_cfg.cpp (hunk 4 above)
Comment (alatarum) [reply src/hw/pcie_cfg.cpp:156]: Right, this is special case and should be handled separately. Will fix.

File: src/hw/pcie_cfg.cpp [hunk 5]
```diff
@@ -83,6 +94,37 @@ bool pcieCfg::addBifurcationConfig(const int& socket,
                           entry("VALUE=%s", slots.c_str()));
     }
 
+    // FIXME BBMC-1020: we hardcode RADUNI address for now since currently we
```
Comment (nest1ing) [inline src/hw/pcie_cfg.cpp]: We agreed do not specify our internal JIRA IDs in public sources.

**Review by AlexanderAmelkin — APPROVED** [review]

**Pull request merged**